design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,123 @@
+# KLIP-30 - Lambda Expressions
+
+**Authors**: derekjn, blueedgenick |
+**Release Target**: N/A |
+**Status**: _In Discussion_ |
+**Discussion**: [#5661](https://github.com/confluentinc/ksql/pull/5661)
+
+**tl;dr**: The introduction of lambda functions would bridge the gap somewhat between builtins and UDFs.
+
+## Motivation and background
+
+While ksqlDB’s UDF interface makes it relatively easy for users to add their own functionality to invoke within their queries, the process of implementing and deploying a UDF is not particularly straightforward. Furthermore, the deployment of UDFs may not even be an option in some environments. Without UDFs, users are at the mercy of ksqlDB’s builtins, which may not always provide enough functionality for them to solve their specific problems. This can be particularly problematic with collections (i.e. `Array` and `Map`), as it can be awkward to work with all of their individual values within a single query. However, we can mitigate this limitation somewhat by empowering users to express user-defined functionality in a way that doesn’t require them to implement full-fledged UDFs.
+
+Introducing lambda functions would enable users to express simple inline functions that can be applied to input values in various ways. For example, lambda functions could be applied to each element of a collection, resulting in a transformed output collection. Lambda functions could also be used to filter the elements of a collection, or reduce a collection down to a single value.
+
+The remainder of this document will propose the addition of lambda functions into ksqlDB’s grammar and execution engine.
+
+## Scope
+
+### What is in scope
+
+* Syntax for defining lambda functions will be proposed
+* Lambda invocation functions will be proposed
+
+## What is not in scope
+
+* Implementation details will not be proposed
+* Level of effort will not be estimated
+
+## Value/return
+
+The introduction of lambda functions will ultimately empower ksqDB users to solve more problems with less effort. Many of the UDFs we’ve seen users write are designed to apply relatively simple functionality to the elements of `Arrays` and `Maps`. Lambda functions are well-suited to solve this class of problems in a very easy and intuitive way.
+
+Lambda functions will also enable some users to use ksqlDB in environments that do not allow for the deployment of UDFs.
+
+## Public APIs
+
+There are two aspects of the user-facing lambda function interface: syntax for describing the actual lambda functions, and a way to specify how to apply these functions to input values.
+
+## Design
+
+### Syntax
+
+It would probably be ideal to use Java-style lambda syntax, but we already use the `->` symbol to represent `Struct` field lookups. Giving `->` a double meaning (if that’s even possible) would create ambiguity and complexity in the grammar/parser. We therefore propose using the `=>` symbol to signify a lambda function:
```

Comment:
IIRC, there was quite a discussion about `.` vs `->` notation to access nested fields. Might be worth to dig out?

Even if overloading `->` might work, it could be hard to read, eg, `transform_array(myArray, e->e->a)` ? Using `e=>e->a` would easier to read. Or would we require parenthesis: `e->(e->a)` ?